kep/108-Memory-bandwidth-and-LLC-aware-filtering-and-scoring/README.md

```diff
@@ -0,0 +1,145 @@
+
+## Design Details
+The filtering and scoring plugin leverages the metrics retrieved from Prometheus to determine the node candidates for running a pod. The metrics considered by the plugin at the moment are:
+
+* For filtering out oversubscribe nodes: 
+
+    * free memory available (GB)
+    * free memory bandwidth (GB/s)
+
+* For scoring:
+
+    * memory bandwidth (GB/s)
+    * memory latency (ns)
+    * Last Level Cache - l3 - Utilization (Bytes)
+    * Last Level Cache contention -- (MPKI - misses per Kilo Instructions)
+    * CPU utilization (%)
+
+the detail about filtering and scoring will be described below. 
+
+. Filtering
+
+This aspect is resposible for filtering out any nodes that would not fit our pod's profile. 
+The pod's profiles are created by analyzing top-down and finding an average scenario, which we can predict will happen on the cluster as well.
+
+In principle, in order to avoid unnecessary oversubscription, we use an *OVERPROVISIONING* factor, by default set to 2. The factor is provided by the plugin configuration file.
+
+Therefore the formulas we use are:
+
+     if (TOTAL_MEMORY_BANDWIDTH - MEMORY_BANDWIDTH_UTILIZED > OVERPROVISIONING * FUNCTION_MEMORY_BANDWIDTH_PROFILE ):
+        return NODE_PASSED_MEMORY_BANDWITH_FILTER
+     if (FREE_MEMORY > OVERPROVISIONING * FUNCTION_MEMORY_UTILIZATION_PROFILE ):
+        return NODE_PASSED_FREE_MEMORY_FILTER
+
+Afterwards, a node that survives both filters passes the filtering.
+
+. Scoring
+
+For every pod each node will be assigned a list of scores based on:
+
+* available resources when sorted by the scoring metrics (see above)
+    * This means that for a pod, all the nodes will be sorted by memory bandwidth, memory latency, cpu, LLC contention and LLC utilization
+    * All these lists will then go on to the next step
+
+
+* the number of counterweights that are applicable to the node
+    * A counterweight is a pod that has already been deployed on a node, but we assume it is still too soon to notice its impact in the measurements we receive from Prometheus. Therefore, we add its profile (memory bandwidth, cpu, and LLC utilization for now) as a counterweight for a *COUNTERWEIGTH_EXPIRATION_POLICY* period of time to the respective node
+
+* ranking strategy when giving point to the sorted nodes
+    * We allow flexible point awarding for our scheduler plugin. For example, our default strategy is to award 10 points to the node ranking first in a certain resource, 5 points to the second place, 1 point to the 3rd place and 0 points to anyone else.
+
+* the pod affinity to the respective resource
```

Comment:
It will be very useful.
Is this a stand-alone plugin or will it be implemented with `available resources when sorted by the scoring metrics` together ?

Reply:
together.